src/inference/tempering.jl

```diff
@@ -0,0 +1,166 @@
+end
+
+
+
+###########################
+#   MODEL
+###########################
+
+struct TemperedEval{T<:Real}
+    model :: DynamicPPL.Model
+    β     :: T
+end
+
+function (te::TemperedEval)(
+    rng::Random.AbstractRNG,
+    model::DynamicPPL.Model,
+    varinfo::DynamicPPL.AbstractVarInfo,
+    sampler::DynamicPPL.AbstractSampler,
+    context::DynamicPPL.AbstractContext,
+    args...
+)
+    minibatchcontext = DynamicPPL.MiniBatchContext(
+        context,
+        te.β
+    )
+    te.model.f(rng, model, varinfo, sampler, minibatchcontext, args...)
+end
+
+"""
+    MCMCTempering.make_tempered_model
+
+Uses a `TemperedEval` to temper a `model`'s `f` by using a `DynamicPPL.MiniBatchContext` to inject `β` as a multiplier on evaluation
+"""
+function MCMCTempering.make_tempered_model(model::DynamicPPL.Model, β::Real)
+    return DynamicPPL.Model(model.name, TemperedEval(model, β), model.args, model.defaults)
+end
+
+
+
+###########################
+#   SWAPPING
+###########################
+
+function MCMCTempering.get_tempered_loglikelihoods_and_params(
+    model::DynamicPPL.Model,
+    sampler::DynamicPPL.Sampler{<:InferenceAlgorithm},
+    states,
+    k::Integer,
+    Δ::Vector{T},
+    Δ_state::Vector{<:Integer}
+) where {T<:AbstractFloat}
+
+    logπk = MCMCTempering.make_tempered_loglikelihood(model, Δ[Δ_state[k]], sampler, get_vi(states[k][2]))
+    logπkp1 = MCMCTempering.make_tempered_loglikelihood(model, Δ[Δ_state[k + 1]], sampler, get_vi(states[k + 1][2]))
+    
+    θk = MCMCTempering.get_params(states[k][2], sampler)
+    θkp1 = MCMCTempering.get_params(states[k + 1][2], sampler)
+    
+    return logπk, logπkp1, θk, θkp1
+end
+
+
+"""
+    MCMCTempering.make_tempered_loglikelihood
+
+Constructs the likelihood density function for a `model` weighted by inverse temperature `β`
+
+# Arguments
+- The `model` in question
+- An inverse temperature `β` with which to weight the density
+- A `sampler` to instantiate a new `VarInfo`
+- Alongside the current `varinfo_init`
+
+## Notes
+- For sake of efficiency, the returned function is closed over an instance of `VarInfo`. This means that you *might* run into some weird behaviour if you call this method sequentially using different types; if that's the case, just generate a new one for each type using `make_`.
+"""
+function MCMCTempering.make_tempered_loglikelihood(
+    model::DynamicPPL.Model,
+    β::T,
+    sampler::DynamicPPL.Sampler,
+    varinfo_init::DynamicPPL.VarInfo
+) where {T<:AbstractFloat}
+    
+    function logπ(z)
+        varinfo = DynamicPPL.VarInfo(varinfo_init, sampler, z)
+        model(varinfo)
+        return DynamicPPL.getlogp(varinfo) * β
+    end
+
+    return logπ
+end
+
+
+"""
+    get_vi
+
+Returns the `VarInfo` given whatever a sampler's second return component is on an `AbstractMCMC.step`, in some cases this is the `VarInfo` itself, in others it must be accessed as a property of the state
+"""
```

Comment:
Also here the style could be fixed.

Reply:
How should I fix the style for this one, is it acceptable to do something like:

```julia
"""
    get_vi(state) / get_vi(vi)

Returns the `VarInfo` given whatever a sampler's second return component is on an `AbstractMCMC.step`,
in some cases this is the `VarInfo` itself, in others it must be accessed as a property of the state.
"""
get_vi(state::Union{HMCState,GibbsState,EmceeState,SMCState}) = state.vi
get_vi(vi::DynamicPPL.VarInfo) = vi
```

Reply:
We follow the Julia standards for documentation and docstrings (or at least we should): https://docs.julialang.org/en/v1/manual/documentation/

In this example maybe:
```suggestion
"""
    get_vi(state)

Return the `VarInfo` object of the `state` of a sampler.

The `state` is the second component returned by `AbstractMCMC.step`. The `VarInfo` object can
be the `state` itself or a property of the `state`.
"""
```